### ISSUE TITLE
Do not require :? for type test patterns

### ISSUE THREAD

**Issue by cartermp** [fsharp/fslang-suggestions#590]
I propose that we make `:?` implicit when pattern matching on subtypes.  That is, the following pattern:

```
:? Type [as identifier]
```

Should be able to be written without the `:?` in all cases:

```
Type [as identifier]
```

Concretely, this means the following would be valid F#:

```fsharp
open System.Windows.Forms

let RegisterControl (control: Control) =
    match control with
    | Button as button -> button.Text <- "Registered."
    | CheckBox as checkbox -> checkbox.Text <- "Registered."
    | _ -> ()
```

This also applies to exceptions:

```fsharp
open System

let divide2 x y =
    try
        Some( x / y )
    with
        | DivideByZeroException as ex -> printfn "Exception! %s " (ex.Message); None
```

My rationale is threefold:

1. I always forget to type `:?`.
2. I think it looks ugly when compared with C# 7, Scala, or Kotlin.
3. I think it helps the problem F# has with existing, skeptical C# developers who see symbols like `:?` and get freaked out.

This a breaking behavioral change.  Consider the following code today in F# (note that `:?` is elided in these patterns):

```fsharp
type foo = int
type bar = float

let f (o: obj) =
    match o with
    | int as i -> printfn "It's an int!"
    | float as f -> printfn "It's a float!"
    
    
let test1 = box 12
let test2 = box 12.0

f test1
f test2
```

Today, this will emit a warning that the second case is never reached, and print the following:

```
It's an int!
It's an int!
```

For this change to make sense, it would have to print:

```
It's an int!
It's a float!
```

I argue that this is not bad for the following reasons:

1. Relying on that behavior is probably very bad.
2. It's sort of nonsensical F# to write today.
3. I believe that the above execution characteristics of the code above violates the principle of least surprise.  This would "fix" that.

## Pros and Cons 

**Pros:**

* Less syntax to remember or type.
* Language looks less "weird" compared with other, similar functional languages.
* Language looks less "weird" to C# developers who point and look at F# and say, "that looks weird to me".
* "Fixes" existing behavior which is likely to be unexpected by newcomers.

**Cons:**

* Potential for name collision on existing DU or exception types, where said types use the same name as an existing class or .NET type, and have no parameters.
* Existing code which a programmer doesn't go in and update would be peppered with `:?` in places.
* It's probably not trivial to implement properly.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S/M

Related suggestions: None that I could find.  I searched for "subtype", "type test", and attempted to search for ":?", but GitHub doesn't support searching by symbols.

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] I or my company would be willing to help implement and/or test this
* [ ] This is not a breaking change to the F# language design

**Comment by matthid** [comment]
> I always forget to type :?.

happened to me as well. But I'm quite a bit more skeptical to this proposal.

>  This is not a breaking change to the F# language design

It is a breaking change (at least in it's current form):

`| Kitty -> printfn "Meow, I'm a kitty!"`

and I think even `| string as s -> printfn "I'm a string!  My content is %s" s` are already valid and assign the matched expression to the `Kitty`, `string` or `s` identifier respectively... 

So at the very least this proposal needs to clarify when it is introducing and when it is considering a type match (When the type is in scope? Some specific types? Based on casing of the expression?)

Does this mean code will break by introducing an `open ...`?

This proposal also might make difficulties for IDE support (ie why suggest a type or open when the code is already valid).

That all said I agree that the `:?` syntax is kind of ugly and alien. But it is as it is, I'd rather stick with it than have multiple syntax elements for the same thing...

**Comment by cartermp** [comment]
> So at the very least this proposal needs to clarify when it is introducing and when it is considering a type match (When the type is in scope? Some specific types? Based on casing of the expression?)

Same rules as with subtypes.

You are correct that you can do some of this just fine today:

```
> let f (o: obj) =                          
-     match o with
-     | string -> printfn "I'm a string!"           
- ;;
val f : o:obj -> unit

> f("hey");;
I'm a string!
val it : unit = ()
```
I've edited the sample above to be a different example.

I brought up this point as a con:

> Name collision on existing DU or exception types, where said types use the same name as an existing class or .NET type, and have no parameters.

Are there other examples where you think this would break existing code?  What scoping issues do you think may occur?

**Comment by smoothdeveloper** [comment]
I find the `:?` / `:?>` syntax logical, based on `:` used for type annotation and the uncertainty of the cast result.

Similarly with `:>` upcast operator.

I'd rather try to explain the difference around `:` symbols, between up and downcast operators (as a type safety feature) and keep the pattern matching syntax as is, which makes it easy to distinguish the kind of match applied.

**Comment by matthid** [comment]
> Same rules as with subtypes.

I don't understand :)

> Are there other examples where you think this would break existing code? What scoping issues do you think may occur?

The question is if we consider it breaking if it currently yields a warning. Because even your new example is valid code as of today, but yields `warning FS0049: Uppercase variable identifiers should not generally be used in patterns, and may indicate a misspelt pattern name.` and ` warning FS0026: This rule will never be matched`

```
open System
let test (pat: obj) =
    match pat with
    | Object as o -> printfn "Object: %O" o
    | String -> printfn "String: %O" String
    | _ -> ()
```

```
> test "test";;
Object: test
val it : unit = ()
```

I'm not saying anyone is using code like the above, but it's definitely (imho) breaking.

**Comment by vasily-kirichenko** [comment]
I think the difference between matching on type and binding to a variable would be too subtle:
```fsharp
type foo = int

// here it's kind of OK because `as` makes it clear that `foo` is a type
match x with
| foo as y -> () // y is int
| bar -> () // it's a variable of type `obj`

// here there is no `as` and the both branches look the same, which is confusing
match x with
| foo -> () // x is int
| bar -> () // it's a variable of type `obj`

```

**Comment by cartermp** [comment]
> I don't understand :)

I'm saying that there should be no differences in semantics of this when compared with existing type test syntax.  That is:

```match x with | :? FooBar as fb-> ...```

Should be semantically equivalent to:

```match x with | FoorBar -> ...```

However, this does bring up the following issue which I've acknowledged as breaking.

```fsharp
type Bananas() = class end
type Test = Bananas | Apples

let f (o: obj) = match o with | Bananas as b -> "It's a bunch of bananas"
```

This currently yields an error:

```
Error FS0001: This expression was expected to have type
    'obj'    
but here has type
    'Test'    
```

Should this continue to break, or should `Bananas` be inferred to be the class rather than the DU?

**Comment by matthid** [comment]
Tbh I think this change is way to dangerous.

Even when we limit this proposal to `as` expressions as @vasily-kirichenko suggests it introduces a inconsistency, because currently you can use `:?` everywhere for example:

`match x with | SomeDuCase(:? Type as t) -> ...` so should we allow the new expression everywhere as well? IMHO that would be even more dangerous... But then again if not, it just is inconsistent (and I don't like that)

**Comment by cartermp** [comment]
That's an interesting case (no pun intended).  In general, yes, I think that allowing the new expression everywhere where it's being used as a pattern would make sense.  I can't think of any new way this could introduce a break than the existing suggestion already.

So to clarify, anywhere where `:? SomeType [as alias]` is used as a pattern should be able to elide the `:?`.  I'll have to think about it some more, but I'm not aware of

And just to clarify, I am **not** suggesting that the following be changed:

```fsharp
let item = box 12
let thing = item :? System.Random
```

Because these are casts instead of patterns.

As for @vasily-kirichenko's point, the following code:

```fsharp
// here there is no `as` and the both branches look the same, which is confusing
match x with
| foo -> () // x is int
| bar -> () // it's a variable of type `obj`
```

So long as  `x` could not be observed as an int outside the scope of the expression associated with `foo`, this doesn't seem strange to me at all.  I think this is also consistent with a `when` expression in Kotlin, and perhaps there are other languages where this is true.

**Comment by cartermp** [comment]
I've updated the suggestion with more information, specifically to include my reasoning for why this "fixes" what I consider to be bad behavior.

**Comment by vasily-kirichenko** [comment]
What about adding a new keyword "is" which is an alias to ":?"? 

```fsharp
let isInt = x is int 

match x with
| is int -> ...
| is string as s -> ...
| DU.Case(is float as f, Bar _, is DateTime) -> ...
```

Maybe we could omit "as", like it's done in C#.

**Comment by matthid** [comment]
So now if Newtonsoft.Json or FSharp.Core introduces a type named "x" all existing code breaks.
I think this is kind of catastrophic now.

The `is` suggestion of @vasily-kirichenko probably would work but I don't see the point. It's exactly the same size as `?:`.
Sorry to be this destructive, so here is possibly another suggestion, we could play with the `|` case syntax:

```
match x with
?| int x -> ...
?| string s -> ...
| DuCase(:? string as s) -> ...
```

> in your example, x is just of the DU type that SomeDuCase of Type is defined in, so there's no pattern matching for that inner bit

That was only an example of what's possible, the `SomeDuCase` could be defined as holding an object `SomeDuCase of obj` for example to make it a pattern match, also it could be an active pattern

Honestly, I still don't think we need any new syntax here at all.

**Comment by cartermp** [comment]
I'm afraid I don't understand your assertion about breaks here @matthid.  Could you explain your scenario where Newtonsoft.Json names a type `x` and that breaks existing code with this suggestion implemented, but doesn't without this suggestion implemented?

**Comment by matthid** [comment]
```fsharp
type x = X of int

let f (o: obj) =
    match o with
    | :? Type as t -> printfn "It's a type!"
    | x -> printfn "Ok everythin else!"
    
    
let test1 = box 12
let test2 = box 12.0

f (box typeof<int>)
f (box "test")
f (box 3)
f (box (X 3))

// currently
//It's a type!
//Ok everythin else!
//Ok everythin else!
//Ok everythin else!

// suggested
//It's a type!
//throws
//throws
//Ok everythin else!
```

**Comment by matthid** [comment]
It doesn't really matter, what I'm saying is you can easily break existing code by introducing types, correct?

**Comment by vasily-kirichenko** [comment]
> The `is` suggestion of [@vasily-kirichenko](https://github.com/vasily-kirichenko) probably would work but I don't see the point. It's exactly the same size as `?:`

`?:` is super ugly (as any non-standard operators, like `:>` and `:?>`), `is` reads nicely.

So as we have `upcast` as a synonym for `:>`, `is` could be a synonym for `:?`.

**Comment by isaacabraham** [comment]
I too am sceptical of this idea. Type test matching is (in my experience at least) not particularly idiomatic in everyday F# (I suppose with the exception of exception handling (no pun intended)); type test matching is not exhaustive compared to other match types for instance, and we have DUs of which all cases are known to the compiler (which e.g. C# does not have).

That being said, that's not necessarily a reason not to improve type test matching where it's easy to do - I'm not sure whether this would be. If you really want to go down the route of "same as C#" then the identifier bit should be implicitly lopped off, too: -

```fsharp
open System.Windows.Forms

let DoStuffWithControl (control: Control) =
    match control with
    | Button -> control.Click() // control is a Button
    | CheckBox -> printfn "control checked is: %b" control.Checked // control is a CheckBox
    | _ -> ()
```

@cartermp I'm not sure if it's worth conflating two orthagonal changes into this single issue - the first seems to me to be about syntax, nothing more. The second seems to be about matching on integers vs floats (I'm not sure how you'd generalise this issue), but that couldn't that be addressed entirely separately from this syntax issue?

**Comment by vasily-kirichenko** [comment]
> If you really want to go down the route of "same as C#" then the identifier bit should be implicitly lopped off, too: -

That's not same as C#, that's same as Kotlin.

**Comment by gerardtoconnor** [comment]
I know there is probably a good reason but why is type inference not used for type matching eg:
```f#
match foo with
| a:string -> printfn "string %s received" a
| _:int -> printfn "received int but value doesnt matter"
| f:float -> printfn "I do care about float values like %f" f
```
this format for me is more consistent with general type inference/declaration patterns elsewhere any may make more sense to newcomers?

**Comment by theprash** [comment]
@gerardtoconnor That would be quite nice syntax if it weren't for the fact that it is already used for a normal type assertion. In your example, if you add some parens, F# currently reads that as assertions that foo is a string, foo is an int and foo is a float. Of course, this is a type error. This type assertion in pattern matches is completely necessary because functions use pattern matching in their parameters, e.g.: `fun (KeyValue(k:int, _)) -> k`.

An alternative would be to use a slightly different operator like `:?`

```fsharp
match foo with
| a :? string -> printfn "string %s received" a
| _ :? int -> printfn "received int but value doesnt matter"
| f :? float -> printfn "I do care about float values like %f" f
```

However, this doesn't give us much more than what we already have (`:? string as a`). It would be really nice if it was in the original language design, but adding it now would contribute to what I think is already a weakness of F#: there are too many ways to do the same thing.

**Comment by gerardtoconnor** [comment]
@theprash I know the default behaviour for type assertions but I was suggesting perhaps a special case in match statements, I have never had to assert types in match case parameters before, perhaps in the match value, so I was wondering if there is any reason a special contextual match type assertion couldn't be built in that acts as a match case for type? In theory being an assertion itself is consistent as being unable to meet the assertion fails case and moves to try next... might be too messy rewiring on match though ... just suggestion though.

**Comment by cartermp** [comment]
@matthid I'm not sure I understand this:

> It doesn't really matter, what I'm saying is you can easily break existing code by introducing types, correct?

If Json.NET or FSharp.Core were to come out with a type called `X` which broke all existing code, this suggestion wouldn't make that any better or worse.  And this suggestion doesn't introduce types.  I'm not sure what your concern here is.

@isaacabraham Regarding this:

> If you really want to go down the route of "same as C#" then the identifier bit should be implicitly lopped off, too

The `[as identifier]` bit means that this is optional.  I should clarify that.  We use this terminology in the F# docs, but it may not be obvious what it means.

> I'm not sure if it's worth conflating two orthagonal changes into this single issue - the first seems to me to be about syntax, nothing more. The second seems to be about matching on integers vs floats (I'm not sure how you'd generalise this issue), but that couldn't that be addressed entirely separately from this syntax issue?

Unfortunately, there's no good way to "unconflate" them.  The second issue isn't about matching on ints and floats.  It's about matching on types in general.  Example:

```fsharp
type C1() = class end
type C2() =
    inherit C1()
type C3() = 
    inherit C1()

type DU = Case1 of C1 | Case2 of C1

let f o =
    match o with
    | Case1 (C2) -> printfn "C2"
    | Case1 (C3) -> printfn "C3"
    | _ -> printfn "neither a Case1 of C2 nor C3";;
```

This will always match the `Case1(C2)` case (and the compiler warning will say as much):

```
> f (Case1(C3()));;
C2
val it : unit = ()

> f (Case1(C1()));;
C2
val it : unit = ()
```

Although the compiler warning will tell me that this behavior will be observed, I still find it strange and unexpected.  I don't think there's a good way to relax usage of `:?` without also dealing with this.

@theprash This is a good point:

> It would be really nice if it was in the original language design, but adding it now would contribute to what I think is already a weakness of F#: there are too many ways to do the same thing.

C# is in a similar place where it's adding features which clean up or replace older ways to do things.  Things like this are also toolable in VS and VSCode, so I could easily envision getting a lightbulb suggesting to remove it.  I think this falls under the category of simplifying and existing pattern, not simply providing a different way to do something.  It's my hypothesis that the large majority of F# developers would choose to elide `:?` in type test patterns if they had the ability to do so.

**Comment by dsyme** [comment]
AFAICS from the original thread on twitter this is about two separate things

1. Make the types of identifiers refined in inner scopes after they have been matched by a corresponding ``:? ty`` 

2. Remove the ``:?`` syntax in type test patterns OR rename to ``is`` 

These are separate.  TBH (1) seems of deeper interest, though I appreciate what's driving (2) as well. I want to focus on (1) here.  With just (1) then this would be allowed:
```
open System.Windows.Forms

let RegisterControl (control: Control) =
    match control with
    | :? Button -> control.Text <- "Registered."
    | :? CheckBox -> control.Text <- "Registered."
    | _ -> ()
```

Both (1) and (2) would be breaking changes though one could reasonably detect the changes for (1)

As I mentioned on twitter, refining the type of identifiers without rebinding doesn't really occur in any ML dialects I know of (well, refinement-type languages like F* may have it, and maybe others, but I'm meaning closer OCaml/SML variations - though I may have missed some). I'm not opposed to it fundamentally.

Besides the breaking change concerns there would also be technical details to consider like whether it applies to tuples (Expectation: yes):
```
    match x,y with 
    | (:? Button), (:? Button) -> ...
    | ...
```
or union/records (expectation: no):
```
    match Pair(x,y) with 
    | Pair((:? Button), (:? Button)) -> are types of x/y refined here?
    | ...
```


As a side note the F# compiler isn't particularly well set up to deal with it internally unless you simply implement it by generating a new Val ``control`` for the inner scope (and be careful about things like this-variable-is-unused warnings).  

Finally, I agree with some above that this issue is, relatively speaking, not as pressing for F# as other languages: in F#, union cases and active patterns are used for most matching. But I can still see why people are asking us to look at (1) and (2)  given the 1:1 code comparisons

**Comment by vasily-kirichenko** [comment]
@dsyme wow. Thanks, I'm surprised :) In Kotlin the feature is called "smart cast" and people _love_ it https://kotlinlang.org/docs/reference/typecasts.html 

About breaking changes in general. I think we cannot drive F# forward if we don't introduce them. What about optional features, like Haskell, for example, has? It allows to turn on some features either by passing command like arguments or by special file-scoped pragmas (`{-# LANGUAGE GADTs #-}`).

**Comment by dsyme** [comment]
> In Kotlin the feature is called "smart cast" and people love it 

I can imagine - though I believe Kotlin people are mostly Java (or other OO) people who think in terms of objects, casts and runtime-types-are-our-tags, much more than is normal in F# programming.   And coming from Java they would feel the need for this urgently.

**Comment by cartermp** [comment]
Thanks for the feedback @dsyme, these are good points.  As for this more philosophical point:

> Finally, I agree with some above that this issue is, relatively speaking, not as pressing for F# as other languages: in F#, union cases and active patterns are used for most matching. As I mentioned [elsewhere](https://github.com/dotnet/csharplang/issues/288#issuecomment-288114545) a strong theme in the F# language design is that it deemphasizes the use of the object-oriented single-inheritance class hierarchy as a way of encoding meta-information in your programming. But I can still see why people are asking us to look at (1) and (2) given the 1:1 code comparisons

I think this is still a good place to look at w.r.t making the .NET interop experience nicer for F#.  Although it's nice to "escape" .NET-isms and put things into DU and Record types, I think there are still enough scenarios where we need to deal with code that makes use of OO-style inheritance.  Ways to be able to use pattern matching here seem worthwhile, as it does help F# feel a bit more consistent/cohesive here.  And now that C#/Kotlin/Swift have this sort of pattern matching, I think it's a good idea to remove as much cruft as possible from accomplishing similar tasks.

In other words, I'd love it if this sort of suggestion could be closed on the basis of "you never interact with objects like this in the wild".  Unfortunately, I believe that it's common enough, and a bit of a sticky space as a whole, that we can work to make it better.  Perhaps it's worth stack ranking pain points here.  I don't think this is as strong as a pain as other places, but I still find it to be useful.  And I think that getting weird-looking symbols out of the way is helpful in convincing skeptical C# dev leads when one of their pesky employees tries to convince them to use F# 😄.

**Comment by dsyme** [comment]
@cartermp TBH I'd probably approve-in-principle a more complete version of (1) today, if it also dealt with the breaking change concerns thoroughly and described the exact behaviour for different kinds of patterns.   
 
(I don't think I'd prioritize my own time to implement it though unless I see particularly painful cases)

(2) I'm less convinced by, and I expect the breaking change concerns to be pretty much impossible to deal with there.  Perhaps better error messages saying ``use :?`` would be enough in that case

**Comment by matthid** [comment]
> If Json.NET or FSharp.Core were to come out with a type called X which broke all existing code, this suggestion wouldn't make that any better or worse. And this suggestion doesn't introduce types. I'm not sure what your concern here is.

This suggestion (or (2) as @dsyme describes it) enables such breaking changes to begin with.

> (2) I'm less convinced by, and I expect the breaking change concerns to be pretty much impossible to deal with there. Perhaps better error messages saying use :? would be enough in that case

sums up what I was trying to say (with examples) from the start :).

Even if we manage to not break any existing code, it just feels too fragile to allow breaking non-related code by introducing a type. Like I said you might not even control that -> "update 3rd party library and all your code breaks" is possible with this "feature". The syntax is just too ambiguous, you can't tell if you create a new symbol or match on a type, depends completely on the context. The suggestion currently doesn't even clarify when we use what, I assume we use type-match as long as we can resolve the symbol to a type and introduce a new symbol otherwise?

Also the reverse is critical: If you rename/remove a type your match starts to hit always (yeah I know that it probably doesn't compile in most cases, but the point is that in some cases this might introduce subtle bugs).

**Comment by cartermp** [comment]
> This suggestion (or (2) as @dsyme describes it) enables such breaking changes to begin with.

I'm picking on this scenario because I'm not convinced that this suggestion makes it any worse than today.  A package author introduce a type which could break people regardless of if this were implemented or not.  Is your argument that this change makes that more likely?

> The syntax is just too ambiguous, you can't tell if you create a new symbol or match on a type, depends completely on the context.

I'm afraid I don't understand this statement.  As I understand it, the scenario is that someone is using some arbitrary identifier that isn't `_` to catch all cases of an `obj` (or all which haven't been matched [as specified in your comment above](https://github.com/fsharp/fslang-suggestions/issues/590#issuecomment-312565826)), and that introducing a new type which happens to match that identifier will result in a break.  Are you referring to a new symbol being created elsewhere?

Because as far as I see it, the above scenario is sort of broken already.  Being able to define `type x = int` and then using `x` as an identifier which will match on any `obj` instance does not feel like correct behavior of the language to me.  That 100% violates my expectation that `x`, which is defined to be an `int` above, is matching on anything.

> The suggestion currently doesn't even clarify when we use what, I assume we use type-match as long as we can resolve the symbol to a type and introduce a new symbol otherwise?

I also don't quite understand what you're asking about here either.  I'm not suggesting that a new symbol be introduced anywhere.  What do you mean by "when we use what"?

**Comment by matthid** [comment]
@cartermp Either my English is just too bad or I don't get it. It seems so obvious to me:

> and that introducing a new type which happens to match that identifier will result in a break.

Yes that is what I'm talking about. I think I should generalize a bit more throughout the language to clarify what I mean:

```fsharp
match t with
| ActivePattern x -> ...
| x -> doSomething x
``` 

According to your suggestion the code above is ambiguous depending on whether type `x` exists or not? Correct?
When type `x` exists it doesn't compile because `x` is undefined on the right side of `->` or changes semantics if `x` is defined outside (even worse)?
When type `x` doesn't exist the second line introduces a new symbol `x` as a "catch-all"/alias of `t` (in this case).

I argue that it should be obvious from the syntax what the code does. Are you suggesting that if the type `x` doesn't exist it wouldn't compile anymore? What about the active pattern?

```fsharp
let (|IsNumeric|_|) x = 
    match x : obj with
    | :? int as i -> Some (box i)
    | :? decimal as i -> Some (box i)
    | :? double as i -> Some (box i)
    | _ -> None
let f = function
   | IsNumeric (x) -> "as"
   | rest -> rest.ToString()
```

When we define `type x = int` what do you expect? I can't tell from your suggestion to be honest. Should `"as"` be returned when the input of `f` is an `int` only? It seems like that is the logical consequence.

> A package author can introduce a type which could break people regardless of if this were implemented or not. 

How? I feel like this particular language suggestion adds another quite subtle way to do it...

> Being able to define `type x = int` and then using `x` as an identifier which will match on any `obj` instance does not feel like correct behavior of the language to me. That 100% violates my expectation that `x`, which is defined to be an int above, is matching on anything.

Well, it's just how it works. But at least it is consistent everywhere. You can even go further than that. When you start using F# you are also wondering why the following doesn't work:

```fsharp
let str = calculate ()
match x with
| str -> // only go here when x = str
| _ -> // x <> str
```

I think that's just part of the learning process and there will always be places like that...

Additionally, I think we need to not only look at `match`, you should keep in mind that pattern match is used throughout the language basically everywhere, for example `function` uses the same syntax:

```fsharp
let (|IsInt|_|) x = 
    match x : obj with
    | :? int as i -> Some i
    | _ -> None
    
let f = function
   | IsInt int -> "as" + int.ToString()
   | rest -> rest.ToString()
```

With `function` you need this value binding feature a lot more because you don't have the initial value to match on.

The same is true for `fun` which basically is a "single" pattern:
```fsharp
let f2 : obj -> string = fun (:? string) -> "string"
```

And even `let`

```fsharp
let (:? string : obj) = box ""
```

I guess this suggestions will give everything different semantics (when a equally named type exists).

So as your suggestion is formulated right now it will "break" variables clashing basically everywhere.
If you limit your suggestion to `match` only, it will simply be a inconsistency in the language. Will you limit to "first"-level matches (as indicated by the `ActivePattern` example above)? -> Even more inconsistent.

(Yes I agree that probably nobody is doing the examples above, but I have seen matches/deconstruction of records, now with your suggestion you might introduce type-matches instead of variable bindings. You need a bit of imagination to see my point - ie. just make the pattern-matches a bit more complex).


Here are just two examples which possibly would break:
- https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core/control.fs#L670 (type `exn` is defined) -> will no longer compile because `error "exn" not defined`
- https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core/control.fs#L1149 (type `cont` is defined, though it is generic) -> will no longer compile probably because `"cont" not defined`.

(if we assume we change every pattern match as I shown above basically every method will break with `ccont` or `econt` parameters because they will now represent matches on the `ccont` and `econt` types and no longer define the `ccont` and `econt` symbols)

Long story short:

>  I expect the breaking change concerns to be pretty much impossible to deal with there.

It's probably just not worth the pain.

Sorry for the long post.

So I hope this can finally clarify my position on this :)

**Comment by Rickasaurus** [comment]
I wouldn't mind something like:

```
match v with
| typeof<int> as i -> ...
| typedefof<List<_>> -> ...
```

It would be super nice to match on generics even if you can't actually bind them out to a reasonable typed value.

**Comment by cartermp** [comment]
Closing as breaking change.
